107-tee-boot-attestation.md

```diff
@@ -0,0 +1,498 @@
+running their OS and applications trusting that they are in-fact running
+confidentially.
+
+[SVSM](https://github.com/coconut-svsm/svsm) is a firmware that runs in CVMs and
+is responsible for performing TEE attestation. SVSM performs TEE attestation to
+unlock the CVM's root disk and establish the initial vTPM state. SVSM will
+perform the initial communication with the Keylime Registrar.
+
+vTPMs will be present to ensure the integrity of software running on a CVM (i.e.
+runtime attestation of CVMs). Furthermore, initial vTPM state will be used to
+derive the key to unlock the CVM root disk. Specifically, encrypted VM images
+will be sealed against the initial vTPM state of the CVM. The initial vTPM state
+key will then be encrypted with a TEE Attestation Key stored in the Keylime
+Registrar. Upon CVM boot, the CVM will have access to the encrypted initial vTPM
+state key. To unlock the root disk, the CVM will need the decrypted initial vTPM
+state key. The only way for the CVM to decrypt its initial vTPM state is to
+attest its TEE evidence with the Registrar. Upon successful TEE attestation, the
+Registrar will decrypt the initial vTPM state and send it back to the CVM. With
+this, initial vTPM state is tied to TEE attestation, and initial vTPM state will
+need to be available to run any workloads on CVMs. Thus, TEE attestation will be
+required for running any OS/applications and initializing initial vTPM state on
+a CVM.
+
+#### Story 1
+
+The use-case of this can be broken up into two main phases, being pre-boot TEE
+attestation and post-boot vTPM attestation.
+
+Pre-boot TEE attestation is performed from SVSM to unlock the CVM's root disk
+and establish initial vTPM state for the machine.
+
+Post-boot vTPM registration is performed by the Keylime agent and no functional
+changes are required. However, the agent enrollment policy will need to account
+for TEE attestation (as will be explained).
+
+Pre-boot Attestation
+--------------------
+
+1. Pre-boot VM Provisioning
+
+As mentioned before, CVM root disks are encrypted, with decryption reliant upon
+successful TEE attestation. In the VM provisioning phase, a user provisions an
+encrypted VM image sealed with initial vTPM state, which is in-turn encrypted
+with a key stored within Keylime. This step involves a new Keylime Registrar
+endpoint to fetch the public components of its "TEE Attestation Key" to then
+encrypt the initial vTPM state.
+
+- User creates an encrypted disk image instance from a plain text image
+  template.
+- User creates initial vTPM state for the CVM.
+- User seals encrypted disk image key against the initial vTPM state.
+- User makes POST request to new Keylime Registrar endpoint to fetch TEE
+  attestation key. Keylime Registrar replies with public components of TEE
+  attestation key.
+- User encrypts initial vTPM state with public TEE Attestation Key.
+- User stores encrypted initial vTPM state along with encrypted root disk.
```

Comment:
There's something not quiet right here.

The vTPM state can't be directly encrypted with the TEE Attestation key. The CVM needs to be able to both encrypt + decrypt the vTPM state, because it needs to persist changes to the state at runtime.

The vTPM state needs to be encrypted with a symmetric key, "vTPM state key". This symmetric key then needs to be encrypted with the TEE attestation key, to form the "Encrypted vTPM state key". The latter then needs to be stored in the attestation server, while the encrypted vTPM state is stored by the virt mgmt app alongside the encrypted disk.